src/Propel/Runtime/ActiveQuery/QueryExecutor/AbstractQueryExecutor.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+/**
+ * MIT License. This file is part of the Propel package.
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Propel\Runtime\ActiveQuery\QueryExecutor;
+
+use Propel\Runtime\ActiveQuery\Criteria;
+use Propel\Runtime\ActiveQuery\SqlBuilder\PreparedStatementDto;
+use Propel\Runtime\Connection\ConnectionInterface;
+use Propel\Runtime\Connection\ConnectionWrapper;
+use Propel\Runtime\Connection\StatementWrapper;
+use Propel\Runtime\Propel;
+use Propel\Runtime\ServiceContainer\ServiceContainerInterface;
+use Throwable;
+
+abstract class AbstractQueryExecutor
+{
+    /**
+     * Indicates if the executor performs writes.
+     * Defaults to true, but is overridden by subclasses.
+     *
+     * @see AbstractQueryExecutor::getConnection()
+     *
+     * @var bool
+     */
+    protected const NEEDS_WRITE_CONNECTION = true;
+
+    /**
+     * @var \Propel\Runtime\Connection\ConnectionInterface
+     */
+    protected $con;
+
+    /**
+     * @var \Propel\Runtime\Adapter\SqlAdapterInterface
+     */
+    protected $adapter;
+
+    /**
+     * @var \Propel\Runtime\ActiveQuery\Criteria
+     */
+    protected $criteria;
+
+    /**
+     * @var \Propel\Runtime\Map\DatabaseMap
+     */
+    protected $dbMap;
+
+    /**
+     * @param \Propel\Runtime\ActiveQuery\Criteria $criteria
+     * @param \Propel\Runtime\Connection\ConnectionInterface|null $con
+     */
+    public function __construct(Criteria $criteria, ?ConnectionInterface $con = null)
```

Comment:
with abstract classes, it might be good to think about whether the constructor should be `final`: given a `class-string<AbstractQueryExecutor> $className`, should users of the Propel library be able to do `new $className($criteria, $connection);`?

If concrete implementing classes (i.e. `extends AbstractQueryExecutor`s) should be able to override the constructor, consider having `AbstractQueryExecutor` implement an `interface ConstructableQueryExecutor { function __construct(Criteria $criteria, ?ConnectionInterface $con = null); }`, so implementers are bound to that interface, too.

Reply:
I am not sure, what's the advantage here?